hazelcast/src/main/java/com/hazelcast/instance/impl/Node.java

```diff
@@ -765,25 +785,63 @@ public class NodeShutdownHookThread extends Thread {
-                    }
+                if (!isRunning()) {
+                    return;
+                }
+                final ClusterTopologyIntent shutdownIntent = clusterTopologyIntentTracker.getClusterTopologyIntent();
+                if (clusterTopologyIntentTracker.isEnabled()
+                        && getNodeExtension().getInternalHotRestartService().isEnabled()
+                        && shutdownIntent != ClusterTopologyIntent.IN_MANAGED_CONTEXT_UNKNOWN
+                        && shutdownIntent != ClusterTopologyIntent.NOT_IN_MANAGED_CONTEXT) {
+                    final ClusterState clusterState = clusterService.getClusterState();
+                    logger.info("Running shutdown hook... Current node state: " + state
+                                + ", detected shutdown intent: " + shutdownIntent
+                                + ", cluster state: " + clusterState);
+                    clusterTopologyIntentTracker.shutdownWithIntent(shutdownIntent);
+                } else {
+                    logger.info("Running shutdown hook... Current node state: " + state);
+                }
+                switch (policy) {
+                    case TERMINATE:
+                        hazelcastInstance.getLifecycleService().terminate();
+                        break;
+                    case GRACEFUL:
+                        hazelcastInstance.getLifecycleService().shutdown();
+                        break;
+                    default:
+                        throw new IllegalArgumentException("Unimplemented shutdown hook policy: " + policy);
                 }
             } catch (Exception e) {
                 logger.warning(e);
             }
         }
     }
 
+    public ClusterTopologyIntent getClusterTopologyIntent() {
+        return clusterTopologyIntentTracker.getClusterTopologyIntent();
+    }
+
+    public void initializeClusterTopologyIntent(ClusterTopologyIntent clusterTopologyIntent) {
+        clusterTopologyIntentTracker.initializeClusterTopologyIntent(clusterTopologyIntent);
+    }
+
+    // true when Hazelcast is running in managed context with a tracker enabled
+    public boolean isClusterStateManagementAutomatic() {
+        return clusterTopologyIntentTracker.isEnabled();
+    }
+
+    /**
+     * When running in managed context with automatic cluster state management,
+     * return true if the current cluster size (as observed by Hazelcast's cluster service)
+     * is same as the cluster specification size (as provided by the runtime environment ie Kubernetes).
+     * @return  {@code true} if running with auto cluster state management and current cluster size is same as
+     *          specified one, otherwise {@code false}.
+     */
+    public boolean isClusterComplete() {
+        return clusterTopologyIntentTracker.isEnabled()
+                && clusterTopologyIntentTracker.getCurrentClusterSpecSize() == clusterService.getSize();
```

Comment:
I am not familiar with conditions for observing split-brain in kubernetes and its resolution (eg I know our 5.1 split-brain improvements did not include anything kubernetes-specific because conditions are different than an on-prem deployment). It should be further investigated, can you please create an issue for that?

Reply:
done https://github.com/hazelcast/hazelcast/issues/22435